tests/pyrog/api/test_attribute.py

```diff
@@ -0,0 +1,76 @@
+import pytest
+from faker import Faker
+
+from django.urls import reverse
+
+faker = Faker()
+
+pytestmark = pytest.mark.django_db
+
+
+@pytest.mark.parametrize(
+    "path, slice_name, definition_id, status_code",
+    [
+        (faker.word(), faker.word(), faker.word(), 201),
+    ],
+)
+def test_create_attribute(
+    api_client,
+    resource,
+    path,
+    slice_name,
+    definition_id,
+    status_code,
+):
+    url = reverse("attributes-list")
+
+    data = {
+        "resource": resource.id,
+        "path": path,
+        "slice_name": slice_name,
+        "definition_id": definition_id,
+    }
+    response = api_client.post(url, data)
+
+    assert response.status_code == status_code
+
+
+def test_retrieve_attribute(api_client, attribute):
+    url = reverse("attributes-detail", kwargs={"pk": attribute.id})
+
+    response = api_client.get(url)
+
+    assert response.status_code == 200
+
+
+def test_list_attributes(api_client, attribute_factory):
+    url = reverse("attributes-list")
+    attribute_factory.create_batch(3)
+
+    response = api_client.get(url)
+
+    assert response.status_code == 200
+    assert len(response.data) == 3
+
+
+@pytest.mark.parametrize(
+    "path, slice_name, definition_id, status_code", [(faker.word(), faker.word(), faker.word(), 200)]
+)
+def test_update_attribute(api_client, attribute, path, slice_name, definition_id, status_code):
+    url = reverse("attributes-detail", kwargs={"pk": attribute.id})
+
+    data = {}
+    for field in ["path", "slice_name", "definition_id"]:
+        if locals()[field]:
+            data[field] = locals()[field]
```

Comment:
faker.word() always return a non-empty string right ?

Reply:
> faker.word() always return a non-empty string right ?

Afaik

> I don't get it, why would any of these variables be undefined (or falsy) in the current scope?

`PATCH` allows partial update. Regardless of the business logic, leaving some fields out of the `data` is perfectly fine.

Currently it's true there is no test case with non specified value for this test (we have to check the business logic).

But the test can be simply extend with for instance :
```
@pytest.mark.parametrize(
    "path, slice_name, definition_id, status_code", 
    [
         (faker.word(), faker.word(), faker.word(), 200),
         (None, None, faker.word(), 200),
    ]
)
```

And it will work bc we check that each field isn't `None`.

NB: maybe sometime we'll have a good reason to update a field to `null`. For instance let's say we create a Column and a Join for that Column. After that we change our mind and want to remove the Join. Either we delete the first Column from the backend, or we update it. For the later we'll need to update the test to make it possible to set the value to `null` in the request.

Reply:
ok understood.
Then why not :
```
for field in [path, slice_name, definition_id]:
    if field:
...
```

Reply:
@simonvadee, because we also need the variable name

But what would you think about parametrizing the whole data object?